src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+        private Storage.Objects storageClient = null;
+        private SimpleInterval currentSubShard;
+        private ContextShard nextOutput = null;
+        private boolean readerClosed = false;
+        private int outputCount;
+
+        public SubdivideAndFillReadsIterator(String bam, AuthHolder auth, int outputShardSize, int margin, final ReadFilter optFilter, ContextShard shard) throws IOException, GeneralSecurityException, ClassNotFoundException {
+            this.bam = bam;
+            this.shard = shard;
+            this.optFilter = optFilter;
+            lastValidPos = shard.interval.getEnd() + margin;
+            firstValidPos = shard.interval.getStart() - margin;
+            ArrayList<SimpleInterval> ints =new ArrayList<>();
+            ints.add(shard.interval);
+            subshards = IntervalUtils.cutToShards(ints, outputShardSize);
+            currentSubShardIndex = 0;
+            currentSubShard = subshards.get(currentSubShardIndex);
+
+            if (BucketUtils.isCloudStorageUrl(bam)) {
+                storageClient = auth.makeStorageClient();
+                reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+            } else if (BucketUtils.isHadoopUrl(bam)) {
+                throw new RuntimeException("Sorry, Hadoop paths aren't yet supported");
+            } else {
+                // read from local file (this only makes sense if every worker sees the same thing, e.g. if we're running locally)
+                reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
+            }
+            query = reader.queryOverlapping(shard.interval.getContig(), shard.interval.getStart(), shard.interval.getEnd());
+
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (null==nextOutput) nextOutput = tryNext();
+            return (null!=nextOutput);
+        }
+
+        @Override
+        public ContextShard next() {
+            if (null==nextOutput) nextOutput = tryNext();
+            if (null==nextOutput) throw new NoSuchElementException();
+            ContextShard ret =  nextOutput;
+            nextOutput = null;
+            return ret;
+        }
+
+        // returns the next shard, if any. Otherwise, closes the reader and returns null.
+        private ContextShard tryNext() {
+            if (readerClosed) return null;
+            ContextShard ret = null;
+            while (query.hasNext()) {
+
+                SAMRecord r = query.next();
+                SAMRecordToGATKReadAdapter g = new SAMRecordToGATKReadAdapter(r);
+                // yes, it'd be a tad faster to check before the wrapping.
+                // But this keeps the code a tad simpler.
+                if (!accept(g, shard.interval)) continue;
+                if (null!=optFilter) {
+                    // skip reads that don't pass the filter
+                    if (!optFilter.test(g)) continue;
+                }
```

Comment:
Can you test `optFilter` within `accept()`? Seems more logical for `accept()` to test all the acceptance criteria.